### PR TITLE
Add EventProxy constructor for pointer to resource.

### PR DESCRIPTION
**Pull request by rchen20**
A compilation error occurred when updating the RAJA Proxy repo to RAJA v0.14.0, where Lulesh 2.0 complains about the lack of an EventProxy constructor. This PR should fix the error below, see related Proxy issue https://github.com/LLNL/RAJAProxies/issues/22.

/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/policy/cuda/forall.hpp(221): error: no instance of constructor "camp::resources::v1::EventProxy<Res>::EventProxy [with Res=camp::resources::v1::Cuda]" matches the argument list 
            argument types are: (camp::resources::v1::Cuda *)
          detected during:
            instantiation of "camp::resources::v1::EventProxy<camp::resources::v1::Cuda> RAJA::policy::cuda::forall_impl(camp::resources::v1::Cuda &, RAJA::policy::cuda::cuda_exec<BlockSize, Async>, Iterable &&, LoopBody &&) [with Iterable=const RAJA::RangeStrideSegment &, LoopBody=lambda [](int)->void &, BlockSize=256UL, Async=true]" 
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/pattern/forall.hpp(562): here 
            instantiation of "camp::resources::v1::EventProxy<Res> RAJA::detail::CallForall::operator()(const T &, ExecPol, Body, Res &) const [with T=RAJA::RangeStrideSegment, ExecPol=Segment_Exec, Body=lambda [](int)->void, Res=camp::resources::v1::Cuda]" 
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/index/IndexSet.hpp(353): here 
            instantiation of "void RAJA::TypedIndexSet<T0, TREST...>::segmentCall(size_t, BODY &&, ARGS &&...) const [with T0=RAJA::RangeStrideSegment, TREST=<>, BODY=RAJA::detail::CallForall, ARGS=<Segment_Exec, const lambda [](int)->void &, camp::resources::v1::Cuda &>]" 
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/index/IndexSet.hpp(349): here
            instantiation of "void RAJA::TypedIndexSet<T0, TREST...>::segmentCall(size_t, BODY &&, ARGS &&...) const [with T0=RAJA::ListSegment, TREST=<RAJA::RangeStrideSegment>, BODY=RAJA::detail::CallForall, ARGS=<Segment_Exec, const lambda [](int)->void &, camp::resources::v1::Cuda &>]"
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/index/IndexSet.hpp(349): here
            instantiation of "void RAJA::TypedIndexSet<T0, TREST...>::segmentCall(size_t, BODY &&, ARGS &&...) const [with T0=RAJA::RangeSegment, TREST=<RAJA::ListSegment, RAJA::RangeStrideSegment>, BODY=RAJA::detail::CallForall, ARGS=<Segment_Exec, const lambda [](int)->void &, camp::resources::v1::Cuda &>]" 
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/pattern/forall.hpp(245): here
            instantiation of "camp::resources::v1::EventProxy<Res> RAJA::wrap::forall(Res &, RAJA::policy::indexset::ExecPolicy<SegmentIterPolicy, SegmentExecPolicy>, const RAJA::TypedIndexSet<SegmentTypes...> &, LoopBody) [with Res=camp::resources::v1::Cuda, SegmentIterPolicy=Segment_Iter, SegmentExecPolicy=Segment_Exec, LoopBody=lambda [](int)->void, SegmentTypes=<RAJA::RangeSegment, RAJA::ListSegment, RAJA::RangeStrideSegment>]"
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/pattern/forall.hpp(358): here
            instantiation of "camp::concepts::enable_if_t<camp::resources::v1::EventProxy<Res>, RAJA::type_traits::is_indexset_policy<ExecutionPolicy>> RAJA::policy_by_value_interface::forall(ExecutionPolicy &&, Res &, IdxSet &&, LoopBody &&) [with ExecutionPolicy=elem_exec_policy, Res=camp::resources::v1::Cuda, IdxSet=LULESH_ISET &, LoopBody=lambda [](int)->void]"
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/pattern/forall.hpp(528): here
            instantiation of "camp::concepts::enable_if_t<camp::resources::v1::EventProxy<Res>, RAJA::type_traits::is_resource<Res>> RAJA::forall<ExecutionPolicy,Res,Args...>(Res &, Args &&...) [with ExecutionPolicy=elem_exec_policy, Res=camp::resources::v1::Cuda, Args=<LULESH_ISET &, lambda [](int)->void>]"
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/include/RAJA/pattern/forall.hpp(523): here
            instantiation of "camp::resources::v1::EventProxy<Res> RAJA::forall<ExecutionPolicy,Args...,Res>(Args &&...) [with ExecutionPolicy=elem_exec_policy, Args=<LULESH_ISET &, lambda [](int)->void>, Res=camp::resources::v1::Cuda]"
/usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/build/lulesh-v2.0/RAJA/lulesh-cuda.cpp(287): here

### CODE REVIEW

File: include/camp/resource.hpp [hunk 1]
```diff
@@ -194,6 +194,7 @@ namespace resources
       EventProxy &operator=(EventProxy const &) = delete;
 
       EventProxy(Res r) : resource_{move(r)} {}
+      EventProxy(Res *r) : resource_{move(*r)} {}
```
Comment (rchen20) [inline include/camp/resource.hpp:197]: @trws This fixes the error I'm getting, but I'm wondering if this is ok? Should I be checking to see if the Res is movable?

Comment (trws) [reply include/camp/resource.hpp:197]: I would think neither of those should have move on them honestly.  That said, we've had some conversations about making events move only, but it shouldn't apply to the resources.  The issue is we need to deal with making sure that resources and events are correctly cleaned up, so I've been trying to root out uses of them by address or reference and remove them.  Are there a lot of uses of resources by pointer in lulesh?

Comment (rchen20) [reply include/camp/resource.hpp:197]: Scratch this - turns out I was accidentally using RAJA v0.13.0 which was calling the `EventProxy` constructor with a reference to the Cuda resource (which has been fixed in v0.14.0). There are no uses of resources by pointer in Lulesh, it's just calling the standard RAJA::forall which creates a Cuda resource internally by default.

Instead, there is another Camp runtime error when using RAJA v0.14.0 with the appropriate Camp v0.2.2:
[chen59@rzansel17:build]$ ./bin/lulesh-v2.0-RAJA-cuda.exe                                                                                                                                                                                                                                                    [30/1195]
Running problem size 45^3 per domain until completion
Num processors: 1
Total number of elements: 91125

To run other sizes, use -s <integer>.
To run a fixed number of iterations, use -i <integer>.
To run a more or less balanced region set, use -b <integer>.
To change the relative costs of regions, use -c <integer>.
To print out progress, use -p
To write an output file for VisIt, use -v
See help (-h) for more options

terminate called after throwing an instance of 'std::runtime_error'
  what():  campCudaErrchk(cudaStreamSynchronize(stream)) an illegal memory access was encountered /usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/tpl/camp/include/camp/resource/cuda.hpp:170
Aborted

**Comment by trws** [comment]
That is very likely to be due to an invalid memory access in lulesh code 
that happens to throw in camp.  Probably related to the allocate 
`MemoryAccess` changes.

On 10 Sep 2021, at 17:03, Robert Chen wrote:

> @rchen20 commented on this pull request.
>
>
>
>> @@ -194,6 +194,7 @@ namespace resources
>        EventProxy &operator=(EventProxy const &) = delete;
>
>        EventProxy(Res r) : resource_{move(r)} {}
> +      EventProxy(Res *r) : resource_{move(*r)} {}
>
> Scratch this - turns out I was accidentally using RAJA v0.13.0 which 
> was calling the `EventProxy` constructor with a reference to the Cuda 
> resource (which has been fixed in v0.14.0). There are no uses of 
> resources by pointer in Lulesh, it's just calling the standard 
> RAJA::forall which creates a Cuda resource internally by default.
>
> Instead, there is another Camp runtime error when using RAJA v0.14.0 
> with the appropriate Camp v0.2.2:
> ***@***.***:build]$ ./bin/lulesh-v2.0-RAJA-cuda.exe              
>                                                                                                                                                                                                                                       [30/1195]
> Running problem size 45^3 per domain until completion
> Num processors: 1
> Total number of elements: 91125
>
> To run other sizes, use -s <integer>.
> To run a fixed number of iterations, use -i <integer>.
> To run a more or less balanced region set, use -b <integer>.
> To change the relative costs of regions, use -c <integer>.
> To print out progress, use -p
> To write an output file for VisIt, use -v
> See help (-h) for more options
>
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  campCudaErrchk(cudaStreamSynchronize(stream)) an illegal 
> memory access was encountered 
> /usr/workspace/wsrzc/chen59/allraja/rajaproxy/rajaproxy_git/tpl/RAJA/tpl/camp/include/camp/resource/cuda.hpp:170
> Aborted
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://urldefense.us/v3/__https://github.com/LLNL/camp/pull/86*discussion_r706525360__;Iw!!G2kpM7uM-TzIFchu!lBMHzspk5c0RrQrHf13RrEy--ULkNDFDQkk9q7sFOEMxFeQzDhtM7jAIHM4Dd4xUDw$

File: include/camp/resource.hpp (hunk 1 above)
Comment (trws) [reply include/camp/resource.hpp:197]: Do you still want to have this constructor or should we close this one out?

Comment (rchen20) [reply include/camp/resource.hpp:197]: Whoops, meant to close this. If need be, I'll make another issue for the cudaStreamSynchronize error.

**Pull request closed without merging**